### PR TITLE
job-info: Add stats for number of jobs in each state

### PR DESCRIPTION
**Pull request by chu11**
Per discussion in #2498, have a way to supply the number of jobs in each of the various job states.  I expose the counters via stats.

The meat of this PR is the job state counting.  How the user gets the data could be something else, but I decided to do it via stats.

### CODE REVIEW

**Comment by grondo** [comment]
what is the difference between `jobs` and `job_states` keys in the `stats` output?
```
$ flux module stats job-info
{
 "lookups": 0,
 "watchers": 6,
 "guest_watchers": 6,
 "jobs": {
  "pending": 3329,
  "running": 32,
  "inactive": 735
 },
 "job_states": {
  "depend": 0,
  "sched": 3329,
  "run": 32,
  "cleanup": 0,
  "inactive": 735
 }
}
```

To me it feels a bit awkward to include low level module stats with these job statistics which are meant for consumption by a high level tool. However, I can't think of any other objections besides aesthetics at this point, so it is probably fine.


One other thought -- for a tool that might want to implement, say, a progress bar based on this output, it might be convenient to include a `total` number of jobs. However, that is redundant information so I'd understand if you want to keep it out of the response.

(e.g. it would make this possible in `jq`:
```
$ echo $(flux module stats job-info | jq '.jobs.inactive / .jobs.total * 100')%
```

**Comment by chu11** [comment]
> what is the difference between jobs and job_states keys in the stats output?

Ahh, I had meant to update that.  `jobs` is the length of the job queues, while `job_states` is .. well job states.  I should probably update `jobs` to `job_queues` or something similar.

> One other thought -- for a tool that might want to implement, say, a progress bar based on this output, it might be convenient to include a total number of jobs. However, that is redundant information so I'd understand if you want to keep it out of the response.

Seems like a good idea.

**Comment by grondo** [comment]
> Ahh, I had meant to update that. jobs is the length of the job queues, while job_states is .. well job states. 

Oh, sorry, that is obvious now. Can the length of the job queues be inferred from the number of jobs in a given state? I.e. will `sched_count` not always equal `zlistx_size (pending)` and `inactive_count` always equal `zlistx_size (inactive)` ? (Sorry in advance for the dumb question)

**Comment by chu11** [comment]
> Oh, sorry, that is obvious now. Can the length of the job queues be inferred from the number of jobs in a given state? I.e. will sched_count not always equal zlistx_size (pending) and inactive_count always equal zlistx_size (inactive) ? (Sorry in advance for the dumb question)

Hmmm.  In the general sense yes, but there is a tiny racy component.

Internally I have a list called `processing`, which will not put a job onto one of the queues until info from the job's eventlog can be retrieved (e.g. userid, t_submit, priority).  B/c I can't sort the information on the queues until I have such information.

So in some respect, the queues have some racy wiggleroom to catch up on until it gets the eventlog information.

But in some general sense, you can always get the "correct" queue length from the job states.

**Comment by grondo** [comment]
Ah, ok. So the list sizes really do seem like an internal module "stats" while the state counts are more of real job statistics, useful for user tools. That seems to make sense.

**Comment by garlick** [comment]
Yeah, I'd suggest breaking out the state histogram to its own query.  The `stats.get` interface tends to get used for module debugging, so it might be tempting to change the returned payload in the future without remembering that, in this case, it's a user facing thing.  Then I'd open the new query to guests.

**Comment by chu11** [comment]
> Yeah, I'd suggest breaking out the state histogram to its own query. The stats.get interface tends to get used for module debugging, so it might be tempting to change the returned payload in the future without remembering that, in this case, it's a user facing thing. Then I'd open the new query to guests.

That's a good point.  I'll do an RPC instead.

**Comment by garlick** [comment]
Another thought:  maybe add `flux job stats` to make this query and dump the results?  I'd be fine with just dumping the json object to stdout for a plumbing command..

**Comment by chu11** [comment]
per discussion above, made the following tweaks

- now grab job stats data via `job-info.job-stats` RPC
- added a "total" field
- add new `flux job stats` command, which just dumps the returned json object directly, e.g.

```
>src/cmd/flux start flux job stats
{"job_states":{"depend":0,"sched":0,"run":0,"cleanup":0,"inactive":0,"total":0}}
```

File: t/t2204-job-info.t [hunk 1]
```diff
@@ -234,6 +234,21 @@ test_expect_success 'flux job list all jobs works' '
         test_cmp list_all_jobids.exp list_all_jobids.out
 '
 
+test_expect_success HAVE_JQ 'job stats lists jobs in correct state (mix)' '
+        depend=`flux job stats | jq .job_states.depend` &&
+        sched=`flux job stats | jq .job_states.sched` &&
+        run=`flux job stats | jq .job_states.run` &&
+        cleanup=`flux job stats | jq .job_states.cleanup` &&
+        inactive=`flux job stats | jq .job_states.inactive` &&
+        total=`flux job stats | jq .job_states.total` &&
```
Comment (grondo) [inline t/t2204-job-info.t]: Just as an FYI, you can test JSON values directly in `jq`, e.g. `flux job stats | jq -e '.job_states.depend == 0`. No need to change anything here though.

Comment (chu11) [reply t/t2204-job-info.t]: oh, haha, I knew that, but I was tweaking the old patch into the new one and just didn't think to tweak in that way.  I'll go ahead and tweak it to use `-e`.

Comment (grondo) [reply t/t2204-job-info.t]: Ah, ok. I just wanted to make you aware if you weren't already since it is useful! :-)

**Comment by chu11** [comment]
tweaked per @grondo comments above

**Comment by grondo** [comment]
Oh, this was fun:
```
$ srun -N93 flux start
$ flux mini run -n1 --dry-run hostname >job.json
$ flux mini run -N1 --dry-run t/ingest/submitbench -r 1024 job.json > submitjob.json
$ t/ingest/submitbench -r 93 submitjob.json
$ while : ; do flux job stats; sleep 0.5; done
```

However, I gave up waiting for 93K jobs to finish and exited my session, then hit an abort from rank 0 flux-broker.

```console
$ exit
lt-flux-broker: src/zlistx.c:360: zlistx_detach: Assertion `self->size == 0' failed.
srun: error: fluke4: task 0: Aborted (core dumped)
$ gdb src/broker/.libs/lt-flux-broker fluke4-flux-broker-0-31882.core 
GNU gdb (GDB) Red Hat Enterprise Linux 7.6.1-115.el7
Copyright (C) 2013 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-redhat-linux-gnu".
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>...
Reading symbols from /g/g0/grondo/git/f.git/src/broker/.libs/lt-flux-broker...done.

warning: core file may not match specified executable file.
[New LWP 32104]
[New LWP 31929]
[New LWP 31928]
[New LWP 31931]
[New LWP 31882]
[New LWP 31990]
[New LWP 32018]
[New LWP 32075]
[New LWP 31930]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Core was generated by `/g/g0/grondo/git/f.git/src/broker/.libs/lt-flux-broker'.
Program terminated with signal 6, Aborted.
#0  0x00002aaaac224377 in __GI_raise (sig=sig@entry=6)
    at ../nptl/sysdeps/unix/sysv/linux/raise.c:55
55	  return INLINE_SYSCALL (tgkill, 3, pid, selftid, sig);
Missing separate debuginfos, use: debuginfo-install czmq-3.0.2-3.el7.x86_64 jansson-2.10-1.el7.x86_64 keyutils-libs-1.5.8-3.el7.x86_64 krb5-libs-1.15.1-37.1chaos.ch6.x86_64 libcom_err-1.42.9-16.el7.x86_64 libgcc-4.8.5-39.el7.x86_64 libselinux-2.5-14.1.el7.x86_64 libsodium-1.0.18-1.el7.x86_64 libstdc++-4.8.5-39.el7.x86_64 lz4-1.7.5-3.el7.x86_64 munge-libs-0.5.13-1.ch6.x86_64 openpgm-5.2.122-2.el7.x86_64 pcre-8.32-17.el7.x86_64 sqlite-3.7.17-8.el7.x86_64 zeromq-4.1.5-3.ch6.x86_64
(gdb) where
#0  0x00002aaaac224377 in __GI_raise (sig=sig@entry=6)
    at ../nptl/sysdeps/unix/sysv/linux/raise.c:55
#1  0x00002aaaac225a68 in __GI_abort () at abort.c:90
#2  0x00002aaaac21d196 in __assert_fail_base (
    fmt=0x2aaaac378c20 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", 
    assertion=assertion@entry=0x2aaaab1c06e9 "self->size == 0", 
    file=file@entry=0x2aaaab1c06c4 "src/zlistx.c", line=line@entry=360, 
    function=function@entry=0x2aaaab1cf0d0 "zlistx_detach") at assert.c:92
#3  0x00002aaaac21d242 in __GI___assert_fail (
    assertion=0x2aaaab1c06e9 "self->size == 0", 
    file=0x2aaaab1c06c4 "src/zlistx.c", line=360, 
    function=0x2aaaab1cf0d0 "zlistx_detach") at assert.c:101
#4  0x00002aaaab17e192 in zlistx_detach () from /lib64/libczmq.so.3
#5  0x00002aaaab17e1e2 in zlistx_delete () from /lib64/libczmq.so.3
#6  0x00002aaaab17e3fa in zlistx_purge () from /lib64/libczmq.so.3
#7  0x00002aaaab17e44e in zlistx_destroy () from /lib64/libczmq.so.3
#8  0x00002aaac13e9661 in job_state_destroy (data=0x2aaad000c590)
    at job_state.c:159
#9  0x00002aaac13e850e in info_ctx_destroy (ctx=0x2aaad000b730)
    at job-info.c:175
#10 0x00002aaac13e86cd in mod_main (h=<optimized out>, argc=<optimized out>, 
    argv=<optimized out>) at job-info.c:218
#11 0x000000000040a5a6 in module_thread (arg=0x67e3e0) at module.c:162
#12 0x00002aaaab678ea5 in start_thread (arg=0x2aaac1826700)
    at pthread_create.c:307
#13 0x00002aaaac2ec8cd in clone ()
    at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
(gdb) 

```

**Comment by chu11** [comment]
> However, I gave up waiting for 93K jobs to finish and exited my session, then hit an abort from rank 0 flux-broker.

Doh!  I assume you hit on fluke?  Will see if I can reproduce on just 1 node.

**Comment by chu11** [comment]
Was finally able to reproduce @grondo's case.  Was able t reproduce on fluke w/ srun on as low as 8 nodes.  Even got it reproducing locally on as local as --size=8 nodes.

The key to reproducing is to wait until submitbench has submitted all 8K jobs before calling exit.  I believe I was calling exit before that point before.

But even at that point, it's not consistent and racy :-(  Running with memcheck_all hasn't caught anything.

**Comment by chu11** [comment]
Running `src/cmd/flux start --size=8 stats.sh`, where stats.sh is

```
#!/bin/sh

rm -f *.json
flux mini run -n1 --dry-run hostname >job.json
flux mini run -N1 --dry-run t/ingest/submitbench -r 1024 job.json > submitjob.json
t/ingest/submitbench -r 8 submitjob.json
while true
do
   flux job stats
   if flux job stats | jq -e ".job_states.total == 8200"; then
       break;
   else
       echo sleeping
   fi
   sleep 5
done

echo sleeping a bit more
sleep 5
```

seems to reproduce fairly reliably.  Putting it into a loop, 7/10 times on opal.

I wanted to run on master, only to stupidly realize `flux job stats` is only in my branch :-)  So I just substituted a "sleep 40" in the end of the script instead.  Hit 7/10 times on master.

So at this point in time, it doesn't appear to be specific to this branch.  Will make an issue.

**Comment by grondo** [comment]
This is cool! I exercised this branch with a little python script to show a progress bar for the instance, including a running average of jobs/s.

demo

Silly script here:

```python
#!/usr/bin/env python
# -*- coding: utf-8 -*-
from __future__ import print_function

import time
import sys
import flux
import collections

def print_progress(fraction, bar_length=72, suffix=""):
    percents = "{0:.1f}".format(100 * fraction)
    filled_length = int(round(bar_length * fraction))
    bar = '█' * filled_length + '-' * (bar_length - filled_length)

    sys.stdout.write('\r|%s| %s%s %s' % (bar, percents, '%', suffix)),

    if fraction == 1.:
        sys.stdout.write('\n')
    sys.stdout.flush()

q = collections.deque(maxlen=10)
h = flux.Flux()
fraction = 0.
last = 0.
suffix = "(0 job/s)"

while fraction < 1.:
    resp = h.rpc ("job-info.job-stats", "{}").get()
    total = float (resp['job_states'][u'total'])
    inactive = float (resp['job_states'][u'inactive'])
    q.append(inactive - last)
    if last > 0:
        suffix = "({0:.1f} job/s)".format((sum(q)/len(q))/0.25)
    last = inactive
    if total > 0.:
        fraction = inactive / total
    print_progress (fraction, bar_length=52, suffix=suffix)
    time.sleep(0.25)
```

**Comment by chu11** [comment]
based on initial results in #2545 and #2547, it appears that the assert @grondo found isn't related to this PR.  Not sure if we want to wait for #2545/#2547 to be fixed first or not.

**Comment by grondo** [comment]
Based on my testing, I think this PR is ready to be merged. I'll review quickly and set merge-when-passing to get the rebase.

Thanks!

**Comment by chu11** [comment]
hit a spurious valgrind fail, restarted that builder

**Pull request merged**